virttest/utils_kernel_module.py

```diff
@@ -0,0 +1,111 @@
+import os
+import logging
+
+from avocado.utils import process
+from avocado.core import exceptions
+
+"""Handle Linux kernel modules"""
+
+
+class KernelModuleHandler(object):
+    """Class handling Linux kernel modules"""
+
+    def __init__(self, module_name):
+        self._module_name = module_name
+        self._was_loaded = True
+        self._config_backup = ""
+        self._backup_config()
+
+    def load_module(self, force=True, params=""):
+        """
+        Load module with given parameters
+
+        :param force: Force load if currently not loaded. Default is True as restore undoes loading.
+        :param params: Parameters to load with, 'key1=param1 key2=param2 ...'
+        """
+
+        if not force and not self._was_loaded:
+            logging.info("Module %s isn't loaded. Set force=True to force loading." % self._module_name)
+        else:
+            cmd = ""
+            if self._was_loaded:
```

Comment:
The same here, this basically results to `rmmod && modprobe` cycle before/after each test, which is a stress test on it's own and might actually influence the testing. How about only doing this in case the specified params differ?

Reply:
Hello @smitterl, what I meant here is that it'd be nice to check the required params from `params` and in case they don't differ from the actual params from `/sys` just leave it as is (without any backup, store or module reload).

Now let's say you want hugepages support turned on (hpage). The current implementation reloads the module no matter what the original setting was the first time it loads and then keeps it enabled until the end of testing where it reverts to the original setting.

The version suggested by me would only get the `/sys` values, checked the `hpage` (that was specified in `params`) is enabled and proceed with no reloads, no state, no cleanup afterwards. On the other hand if the `hpage` was off, it'd have reload the module for each test that asks for it.

So to wrap it up, my solution should be way simpler, should result in faster execution when no changes are required but would result in reload for each test that requires different params (that especially in CI should be eliminated by custom setup). Your solution currently adds one reload that can be eliminated but more importantly adds the state and potentially corner-cases that might be problematic to spot and debug.

PS: I still haven't reviewed the whole code so maybe it's cleverer than I thought, but so far I don't see any checks for the expected and actual setting...

Reply:
PPS: By the potential corner-cases I mean multiple tests requiring different params etc. I haven't checked the code yet but I guess it might become problematic, which is why I'd prefer stateless implementation... (not sure what params you have in mind, I'm only using nested and hpage and don't combine their execution so setting it in advance)

Reply:
@ldoktor As pointed out by you, I think it's a good idea to 
1. check if requested parameters are already set (by comparing with `'cat /sys/module/X/paramters/*'`) - and skip if so
2. add additionally a log to reflect the full module parameters list if the module has been loaded - that is, if not skipping

By adding 1. to the current stateful implementation, we avoid the additional reload you mention. By adding 2. we make sure that full debugging is possible.

The original use case motivating this change is simply to activate hugepage support on s390x which isn't available in our test environments' kernel by default. I have no use cases where several test cases would require different module parameters.

I don't see further advantage of having a stateless implementation as originally proposed: if huge page support is not available by default, as in the original use case, and if the support should be active in each test in a sequence, the module load and restore commands are called only once per sequence which still reduces stress on the system and test execution times.

Therefore, my proposal would be to add 1. and 2. to the current implementation. Please, let me know.